url400631cdd0_ludaiqian_appliedxml_tgz-pJ8-test_simples_TestJ8/src/com/google/gsoncode/internal/LinkedTreeMap.java

```diff
@@ -0,0 +1,494 @@
+    return null;
+  }
+
+  private void rebalanceAfterInsert(TreeNode<K, V> x) {
+    x.color = RED;
+
+    while (x != null && x != root && x.parent.color == RED) {
+      if (x.parent == leftOf(parentOf(parentOf(x)))) {
+        TreeNode<K, V> y = rightOf(parentOf(parentOf(x)));
+        if (colorOf(y) == RED) {
+          setColor(parentOf(x), BLACK);
+          setColor(y, BLACK);
+          setColor(parentOf(parentOf(x)), RED);
+          x = parentOf(parentOf(x));
+        } else {
+          if (x == rightOf(parentOf(x))) {
+            x= parentOf(x);
+            rotateLeft(x);
+          }
+          setColor(parentOf(x), BLACK);
+          setColor(parentOf(parentOf(x)), RED);
+          rotateRight(parentOf(parentOf(x)));
+        }
+      } else {
+        TreeNode<K, V> y = leftOf(parentOf(parentOf(x)));
+        if (colorOf(y) == RED) {
+          setColor(parentOf(x), BLACK);
+          setColor(y, BLACK);
+          setColor(parentOf(parentOf(x)), RED);
+          x = parentOf(parentOf(x));
+        } else {
+          if (x == leftOf(parentOf(x))) {
+            x = parentOf(x);
+            rotateRight(x);
+          }
+          setColor(parentOf(x), BLACK);
+          setColor(parentOf(parentOf(x)), RED);
+          rotateLeft(parentOf(parentOf(x)));
+        }
+      }
+    }
+    root.color = BLACK;
+  }
+
+  private static <K extends Comparable<K>, V> TreeNode<K, V> parentOf(TreeNode<K, V> e) {
+    return (e != null ? e.parent : null);
+  }
+
+  private static <K extends Comparable<K>, V> boolean colorOf(TreeNode<K, V> e) {
+    return (e != null ? e.color : BLACK);
+  }
+
+  private static <K extends Comparable<K>, V> TreeNode<K, V> leftOf(TreeNode<K, V> e) {
+    return (e != null ? e.left : null);
+  }
+
+  private static <K extends Comparable<K>, V> TreeNode<K, V> rightOf(TreeNode<K, V> e) {
+    return (e != null ? e.right : null);
+  }
+
+  private static <K extends Comparable<K>, V> void setColor(TreeNode<K, V> e, boolean c) {
+    if (e != null){
+      e.color = c;
+    }
+  }
+
+  private static <K extends Comparable<K>, V> TreeNode<K, V> successor(TreeNode<K, V> t) {
+    if (t == null) {
+      return null;
+    } else if (t.right != null) {
+      TreeNode<K, V> p = t.right;
+      while (p.left != null) {
+        p = p.left;
+      }
+      return p;
+    } else {
+      TreeNode<K, V> p = t.parent;
+      TreeNode<K, V> ch = t;
+      while (p != null && ch == p.right) {
+        ch = p;
+        p = p.parent;
+      }
+      return p;
+    }
+  }
+
+  private void rotateLeft(TreeNode<K, V> p) {
+    if (p != null) {
+      TreeNode<K, V> r = p.right;
+      p.right = r.left;
+      if (r.left != null) {
+        r.left.parent = p;
+      }
+      r.parent = p.parent;
+      if (p.parent == null) {
+        root = r;
+      } else if (p.parent.left == p) {
+        p.parent.left = r;
+      } else {
+        p.parent.right = r;
+      }
+      r.left = p;
+      p.parent = r;
+    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 195:211, 215:231.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.